### ISSUE TITLE
FLTK 1.4 Window Scaling does not work correctly when window is maximized or full screen

### ISSUE THREAD

**Issue by dannye** [fltk/fltk#919]
**Describe the bug**
On Windows and Mac, scaling the window in or out has unpredictable effects if the window is maximized or full screen. It works decently on Linux.

**To Reproduce**
You can use almost any of the demo programs to see this issue. In my case, I used **test/tree**.

1. Compile the default test programs
2. Run test/tree
3. Maximize or full screen the window
4. Scale the window in or out
5. Trigger some redrawing by interacting with some widgets
6. See that the screen is painted unexpectedly

**Expected behavior**
The "solution" is probably a bit debatable.

My two ideas on what I would expect to happen are:

1. If the window is maximized or full screen, the window should be rescaled and resized _simultaneously_ in order to continue using up the whole screen. (This is what happens on Linux.)
2. Or, the scale in/out operation should be blocked/ignored while the window is maximized or full screen.

**Videos**
Windows:

https://github.com/fltk/fltk/assets/5231310/d7768cd7-5061-4c33-8650-cd6b0005208f

https://github.com/fltk/fltk/assets/5231310/7d29e0f5-0c35-4e65-8ae8-3eb11db26a28

^Another issue specifically on Windows: After maximizing, scaling, and then restoring the window, the window will inexplicably and temporarily reset the scale, usually after dragging the window.

Mac:

https://github.com/fltk/fltk/assets/5231310/68c42ebd-ba6d-45da-8055-24f222f2e4ea

Linux:

https://github.com/fltk/fltk/assets/5231310/cce91b2e-c8ea-4270-bc8e-61d933f3c2a6

**FLTK Version**
 - Version: 1.4.0
 - Downloaded and self-built from: Git
 - If from Git, branch: master
 - If from Git, commit: f288aea

**FLTK Configure / Build Options**
Windows: VS Cmake tools
Mac/Linux: `make`

**Operating System / Platform:**
 - OS: Windows
 - OS Version: Windows 11
 - Processor if relevant: Intel
 
--

 - OS: macOS
 - OS Version: macOS 11.4
 - Processor if relevant: Intel
 
--

 - OS: Linux
 - OS Version: PeppermintOS 10
 - Processor if relevant: Intel

**Comment by Albrecht-S** [comment]
@dannye Thanks for reporting this issue, for testing on all platforms, and for the videos.

My gut feeling is that FLTK's scaling should not be applied (i.e. the keyboard shortcuts should be ignored) while the window is in fullscreen and/or maximized mode. Scaling could very likely result in unpredictable results (as we can see in the posted videos). OTOH, if "scaling" would be disabled, this would always keep a well-defined status.

@ManoloFLTK, @MatthiasWM What are your opinions?

**Comment by dannye** [comment]
It's also worth noting that the previous videos only demonstrated the following scenarios:
* Windows: maximized
* Mac: full screen
* Linux: maximized

But the following scenarios also exist:
* Windows: full screen
* Linux: full screen

You can apply this patch to the auto-generated tree.cxx to turn the D1 button into a full screen toggle and the D2 button into a maximize toggle:

```diff
diff --git a/test/tree.cxx b/test/tree.cxx
index 451ab5982..d2136d900 100644
--- a/test/tree.cxx
+++ b/test/tree.cxx
@@ -22,6 +22,26 @@ const char* reason_as_name(Fl_Tree_Reason reason) {
       }
 }
 
+Fl_Double_Window* window = (Fl_Double_Window*)0;
+
+void Button_CB_D1(Fl_Widget* w, void* data) {
+    if (window->fullscreen_active()) {
+        window->fullscreen_off();
+    }
+    else {
+        window->fullscreen();
+    }
+}
+
+void Button_CB_D2(Fl_Widget* w, void* data) {
+    if (window->maximize_active()) {
+        window->un_maximize();
+    }
+    else {
+        window->maximize();
+    }
+}
+
 void Button_CB(Fl_Widget*w, void*data) {
   // Each push changes height so we can test 'Item h() from widget'
   //    Cycle through larger sizes until 50, then wrap to normal size.
@@ -185,11 +205,11 @@ void RebuildTree() {
                 grp->begin();
                   Fl_Button *abut = new Fl_Button(grp->x()+0 ,grp->y()+2,65,15,"D1");
                   abut->labelsize(10);
-                  abut->callback(Button_CB);
+                  abut->callback(Button_CB_D1);
                   abut->tooltip(tipmsg);
                   Fl_Button *bbut = new Fl_Button(grp->x()+75,grp->y()+2,65,15,"D2");
                   bbut->labelsize(10);
-                  bbut->callback(Button_CB);
+                  bbut->callback(Button_CB_D2);
                   bbut->tooltip(tipmsg);
 
                 grp->end();
@@ -327,8 +347,6 @@ Fl_Color GetSelectedItemBGColor() {
   return(tree->item_labelbgcolor());
 }
 
-Fl_Double_Window *window=(Fl_Double_Window *)0;
-
 Fl_Tree *tree=(Fl_Tree *)0;
 
 static void cb_tree(Fl_Tree*, void*) {
```

I'd say the behavior of Windows + full screened scaling is not great. It's as if it tries to rescale+resize (like on Linux), but it doesn't always preserve the aspect ratio, and doesn't always keep the full screened window at the exact top-left of the monitor which it is full screened to.

Similarly, Linux + full screen scaling messes up the aspect ratio and sends the black terminal off screen below the monitor in test/tree.

`window->fullscreen()` on Mac doesn't really full screen the window correctly in general so it isn't worth testing too much.
I've always used a workaround where I never call `window->fullscreen()` on Mac. Maybe I will submit an issue for this soon.

**Comment by erco77** [comment]
At first glance I kinda agree with Albrecht about scaling should maybe be disabled in full screen mode. At least until some reasonable behavior can be implemented for scaling within a fixed screen size.

The only reasonable behavior I can think of for a "scale up" in fullscreen mode is to first "resize smaller" (without going smaller than window size minimums), then a "scale up" applied to that smaller window to fill the physical screen, assuming it fits. If it doesn't fit, the scale would be ignored.

I suppose Fl_Window size maximums would also have to be taken into account, along with whatever resizable() behavior is configured for the window.

Anyway, I'll bet there's devils in the details, and if that's prohibitive, simply disabling scaling in fullscreen might be best.
Curious to see Manolo + Matt's opinion on this, as they've actually been doing backend stuff.. I'm just speculating from the sidelines.

**Comment by Albrecht-S** [comment]
@erco77 Thanks for your comment, and I apologize that I forgot you in the list of devs I asked for comments.
Greg wrote:
> I kinda agree with Albrecht about scaling should maybe be disabled in full screen mode. At least until some reasonable behavior can be defined ... I'll bet there's devils in the details, and if that's prohibitive, simply disabling scaling in fullscreen might be best.

I forgot to make a clear proposal (potential roadmap), like this:
1. Disable the shortcuts in full-screen mode for 1.4.0 (or 1.4.x altogether).
2. Discuss and define a reasonable behavior how scaling in fullscreen mode **should** work (if at all).
3. Implement whatever we defined in 2. in "next gen FLTK" (1.5.0).

**_Reasoning:_** I'm afraid that this task (2.+3.) is not easy and can be very problematic for unknown apps that implement their own resizing behavior in non-fullscreen mode. I wouldn't want to open yet another can of worms in FLTK 1.4.0 before the release. Disabling the function should be easy to do though.

**Comment by ManoloFLTK** [comment]
Obviously, it makes no sense to scale up or down a fullscreen or maximized window.

**Comment by Albrecht-S** [comment]
> Obviously, it makes no sense to scale up or down a fullscreen or maximized window.

From a user's point of view it **could** make sense. Imagine you have a small window like test/hello, then make it fullscreen. Now you notice that the font is too small and type ctrl/+. Strange things happen...

OTOH, if you had pressed ctrl/+ once or multiple times **before** you made the window fullscreen, the result would be fine. Consequently the user would have to leave fullscreen mode, type ctrl/+, and then go back to fullscreen.

What we're discussing here is basically what happens if the user does not switch back to normal window mode before they press ctrl/+, and the proposal for 1.4.0 was to disable the shortcuts although a better implementation **might** be possible - but too complicated to decide for 1.4.0, and thus leaving it to 1.5.0. Again, if and only if a useful algorithm could be found for this special case.

**Comment by ManoloFLTK** [comment]
@dannye Please, close issue if you agree it's fixed with the last commit as planned above.

**Comment by Albrecht-S** [comment]
> ... it is still possible to scale a maximized or full screen window, just with more steps.

@dannye What do you mean with this statement? If you mean that a user would have to leave fullscreen mode, then scale (ctrl/+/-) and then go back to fullscreen, then yes, this would be true. There's no reason to prevent this scenario.

Your issue is about inconsistent and obviously unexpected or undefined behavior. My goal in this discussion is to make FLTK **_behave consistently_** and to **_avoid undefined behavior_**. If we **could** find an algorithm that lets us scale the window contents while the window is in fullscreen mode, then this would be fine. However, my take on this is that we can't do that because application specific window resizing could not be "simulated" while the window is in fullscreen mode.

Since we probably can't find a consistent way to do what users would expect (I can't imagine what this would be), it's IMHO the best choice to disable scaling while in fullscreen mode. This doesn't prevent users from going back to window mode, resizing the window if necessary (by dragging the border), then scaling it to another zoom factor, and eventually switch back to fullscreen. Yes, admittedly more steps, but hopefully consistent and expected behavior.

**Comment by dannye** [comment]
> What do you mean with this statement?

I'm talking about an FLTK application that has 2 or more windows. If one window is full screen and the other is not, **and the non full screen window has focus** and the window scaling shortcut is pressed, then based on the commit message in commit 41740ca both windows will be scaled (all windows in a single FLTK application share a single global scaling factor), including the one that is **presently** full screen -- which is exactly what we're talking about blocking.

> My goal in this discussion is to make FLTK **_behave consistently_** and to **_avoid undefined behavior_**.

Of course that is my only goal as well. We are in total agreement.

**Comment by dannye** [comment]
This is what I am describing:

https://github.com/fltk/fltk/assets/5231310/016e9ff1-8b33-471f-bdf1-38dcabe24935

**Comment by ManoloFLTK** [comment]
Commit mentioned above should fix that other scenario.

**Comment by Albrecht-S** [comment]
> I'm talking about an FLTK application that has 2 or more windows. If one window is full screen and the other is not,

OK, I see. FYI: only windows on the same screen share the same scaling factor, but anyway. Your point is valid, and Manolo fixed it by checking all windows on the same screen for fullscreen mode.

Although this may be unexpected (Q: "why does my window not scale?" - A: "because another window on the same screen is in fullscreeen mode.") it's a compromise and I agree with this solution for 1.4. Maybe we can do better in 1.5.

**Comment by dannye** [comment]
> FYI: only windows on the same screen share the same scaling factor, but anyway.

This is empirically not true on Mac. The _only_ reason I put the two windows on the same screen is because Apple QuickTime can only record one screen at a time. If I perform the exact same actions but leave the smaller window on the other monitor, scaling the smaller window simultaneously scales the larger window. This is in fact how I originally observed the bug yesterday, I just couldn't record it.

But now that you mention this, I do see the behavior you're talking about on Windows. I have noticed it in the past but thought it was a bug, because it was so disorienting when I'd drag a window from one screen to another and then its scale would suddenly change 1 full second later. This is kind of shocking for single-window applications IMO, especially when the two monitors have the same resolution.

**Comment by dannye** [comment]
This is what I'm referring to in case it isn't clear:

https://github.com/fltk/fltk/assets/5231310/bd34cc3b-9fe9-476e-a9b1-1c985c7a7542

**Comment by dannye** [comment]
This is actually more problematic on Windows. All that's required is to maximize the window within roughly 1 second of moving the window to a new monitor, which is extremely easy on Windows:

https://github.com/fltk/fltk/assets/5231310/c1664737-e5a0-4646-bed8-bc2dafb68e82

This was a full rebuild with commit 746cbf8.

**Comment by ManoloFLTK** [comment]
Under Windows, a desktop can mix a regular-DPI display and a HighDPI-display. That's the reason why the scaling factor has been made independent per screen so that if the scaling factor, especially that one under control of the Settings Windows app, is set adequately for each screen, an FLTK window can be moved from one such display to the other while keeping its physical size.

Under macOS, systems have typically either regular DPI or a retina display which has double DPI and macOS automatically draws with 2 pixels per graphics unit on retina displays, so the default has been made to have the same scale factor for all displays.

**Comment by Albrecht-S** [comment]
> Under macOS, systems have typically either regular DPI or a retina display which has double DPI and macOS automatically draws with 2 pixels per graphics unit on retina displays, so the default has been made to have the same scale factor for all displays.

@manolo I didn't know that. What does "typically" mean, and what consequences does this have? I have only one MacBook Air with its built-in Retina display. But what if I attach a "normal" monitor with any screen resolution and DPI? Would this be "supported", or what effects would I see? Wouldn't it be better to have the same freedom WRT monitors and DPI we have on other platforms? I'm wondering...

**Comment by Albrecht-S** [comment]
@dannye As I wrote above I didn't know that scaling factors are the same on different screens on macOS. I knew about the adaptation made for different DPI on different monitors on other OS's. I also seem to remember that the small delay before the window is rescaled was for some reason. However, the entire DPI adaptation made for different monitors is to accomodate for the "reality" and, as Manolo wrote, to make it possible to drag a window from one screen to another one and keeping its size, provided that the screen DPI is set up correctly (e.g. in system settings). There's obviously nothing we can do if the user does something to prevent the automatism. You could argue that the delay is too long though, but I can't remember why that delay was implemented. @ManoloFLTK would likely know what the reasoning was.

**Comment by dannye** [comment]
> There's obviously nothing we can do if the user does something to prevent the automatism.

I'm not sure what you mean by this. The current bug that I am showcasing in the latest video is simply the result of _very ordinary_ interactions with a window on the Windows OS that exposes a bug with FLTK. Do you disagree?

**Comment by Albrecht-S** [comment]
@dannye Sorry, I can't really see what's happening on your video, it's too large for my monitor, or the details are too small for my eyes. What I was referring to was:

> All that's required is to maximize the window within roughly 1 second of moving the window to a new monitor, which is extremely easy on Windows

That I don't call "ordinary interactions".

But I don't want to argue with you. All this scaling stuff is not my part, and I don't feel comfortable commenting on this. I only wanted to help but maybe I shouldn't have tried. Manolo knows more about this and we should wait for his comments.

**Comment by dannye** [comment]
> > All that's required is to maximize the window within roughly 1 second of moving the window to a new monitor, which is extremely easy on Windows
> 
> That I don't call "ordinary interactions".

I certainly don't want to argue unproductively either, but I still think we are at a total misunderstanding. So I just want to say this to clarify.
I'm just talking about a single drag-n-drop action -- dragging a window from one screen into the top edge of another screen. I do this probably a hundred times a day. This currently corrupts the window in FLTK 1.4 if the two screens have different scaling factors.

> Sorry, I can't really see what's happening on your video, it's too large for my monitor, or the details are too small for my eyes.

The second half of [this video](https://github.com/fltk/fltk/issues/919#issuecomment-1966708035) shows the window corruption very clearly. I realize the details in the first half of the video are very tiny and I apologize about that.

**Comment by ManoloFLTK** [comment]
> The second half of https://github.com/fltk/fltk/issues/919#issuecomment-1966708035 shows the window corruption very clearly. I realize the details in the first half of the video are very tiny and I apologize about that.

Sorry, I don't see in neither halves of that video what is done, what is expected, and what is obtained. Probably a word description of a failing scenario, preferentially with a simple app (FLTK's editor for instance), would be useful.

**Comment by ManoloFLTK** [comment]
About macOS having a single scaling factor for all its displays whereas other OSes have one for each display.

macOS supports retina displays by itself, without FLTK having to change anything : the system considers that 1 drawing unit corresponds to 1 pixel on a standard display and to 2 pixels on a retina display, FLTK doesn't even need to be aware of that. Remarkably, this makes that windows of an FLTK 1.3 app can be moved from a retina to a regular display and keep their physical size. FLTK 1.4 brings nothing to macOS apps in terms of HighDPI support. I've been working for years with a retina laptop attached to a cheap external display, and all FLTK apps (1.3 or 1.4) are perfectly happy. macOS automagically detects which display is HighRes and which is not and draws FLTK windows accordingly. Given this, it makes sense, I believe, to have a single scaling factor for all displays of a macOS system: if one window content appears unreadable on one display, it has the same size on the other display so will be equally unreadable on the other display.

Windows supports HighDPI displays in a different way. Its Settings app allows to change each display's scaling by various quantities (125%, 150%, 200%,…). Native Windows apps using modern Windows-only GUI API query the scaling value of the display on which a window is mapped and scale the window accordingly. The physical size of a given window is therefore susceptible to change when the window is moved between screens or when a screen is re-scaled with the Settings app. Given this, it makes sense for FLTK to have as default scaling factor of each display the value given by Settings for that display. FLTK 1.4 windows moved across displays behave as modern Windows-only apps following each display's scaling value. In contrast, FLTK 1.3 apps are DPI unaware applications. Citing MS documentation: "They render at a fixed DPI value of 96 (100%). Whenever these applications are run on a screen with a display scale greater than 96 DPI, Windows will stretch the application bitmap to the expected physical size. This results in the application appearing blurry". FLTK 1.4 allows its client apps to draw using the full display's resolution also on HighDPI displays.
 
Wayland uses some combination of the Windows and macOS approaches to support HighDPI. Wayland gives, via the desktop's Settings app, an integer-valued scale factor for any display. That is 1 for regular displays and 2 or possibly 3 for HighDPI ones. Any Wayland window reads the factor value, n, for the display where it's mapped and draws following the rule "1 drawing unit = n pixels". KDE and very recent Gnome versions also support fractional scaling values. They are obtained combining 2 levels : 1) at the Wayland client level, everything works using the smallest integer value larger than the display's fractional scale value (that is, 150% gives n = 2); 2) at the Wayland compositor level, the pixels received from the client are mapped on the display shrunk to the desired fractional scaling.

X11 was born much before HighDPI displays existed. My current understanding is that X11 gives one scaling factor for its desktop, obtained dividing the value of the X11 resource Xft.dpi by 96. FLTK uses that as starting scaling factor for all displays. If the desktop mixes displays with diverse DPIs, each display can be re-scaled independently by Ctrl/+/-/ to compensate for DPI differences.

**Comment by Albrecht-S** [comment]
Thanks, @ManoloFLTK, for this detailed explanation. I didn't know about the "macOS magic" and I believe you that this is alright then. I can't test on macOS with external display anyway, hence I'm fine with this.

@dannye Please see Manolo's comment above. I'm leaving further discussion and resolution of this issue to him, I'm too busy with other FLTK stuff and I can't help much anyway.

**Comment by ManoloFLTK** [comment]
About the 1-sec delay under Windows before an FLTK window moved to a new display adopts that display's scaling factor: I remember that a couple of years ago, there were problems without delay, but now I tested under Windows 11 with no delay at all and find that FLTK windows change display and resize fluidly just as when I move a Notepad++ window. @dannye's videos suggest Windows 10 runs. May be we could just remove the delay altogether? That would require testings on Windows 10 or earlier I can't do.

**Comment by dannye** [comment]
> > The second half of [#919 (comment)](https://github.com/fltk/fltk/issues/919#issuecomment-1966708035) shows the window corruption very clearly. I realize the details in the first half of the video are very tiny and I apologize about that.
> 
> Sorry, I don't see in neither halves of that video what is done, what is expected, and what is obtained. Probably a word description of a failing scenario, preferentially with a simple app (FLTK's editor for instance), would be useful.

1. Open test/tree.exe
2. Scale window up or down
3. Move window to new screen and maximize simultaneously by dragging and dropping against the top edge of the new screen

https://github.com/fltk/fltk/assets/5231310/5a6fdcb4-2772-4adf-8a9e-5642f0683600

This is <ins>exactly the same</ins> as the OP: FLTK does not correctly rescale windows that are presently maximized or full screen.

And just like [this comment](https://github.com/fltk/fltk/issues/919#issuecomment-1965528202) where I am pointing out yet another simple circumstance where FLTK can be "tricked" into rescaling a maximized or full screen window, I am doing exactly the same thing here. I am pointing out another very simple circumstance where FLTK can be tricked into rescaling a window which is presently maximized, which will never produce the expected result. The corruption is exactly the same as all previous examples.

**Comment by dannye** [comment]
Here is why I think this bug is so non-trivial to fix: The "only clean fix" I can think of is not good enough. It is not possible to guarantee that a window is rescaled by FLTK as it is dragged to a new screen before that window becomes maximized.

The reason for this is because windows can be sent to another screen with the keyboard shortcuts Shift+Win+Left and Shift+Win+Right.

See:

https://github.com/fltk/fltk/assets/5231310/a0f032e3-dcdb-4a69-86e6-17f99d3aa728

FLTK has no opportunity to rescale this window because it is already maximized on one screen and remains maximized while it is sent to another screen with a different scaling factor.

Therefore, I think the combination of the two facts (1) Different screens can have different scaling factors and (2) All windows on the same screen share a scaling factor is untenable (unless the potential algorithm for correctly scaling maximized/full screen windows is implemented).

IMO, without knowing all the details, I think Windows should behave more like Mac for the time being. An FLTK application should have a single global scaling factor shared by all screens at all times. This is the only compromise I can think of that will work in the meantime until the potential algorithm is implemented.

**Comment by dannye** [comment]
I can think of one other possible solution for now:

Any time that FLTK is about to attempt to rescale a maximized or full screen window, it can "simulate" this by restoring the window (un maximizing or un full screening), rescaling the window, and then re maximizing or re full screening the window. This may be clunky and may come with its own issues, but just putting the option out there. Perhaps it's possible to simulate these calculations without actually performing all these actions on the window from the OS's perspective, I dunno.

**Comment by Albrecht-S** [comment]
@ManoloFLTK I could maybe test on an older system with Windows 10 and two screens. However, doing this would be a time consuming process, hence I would only want to do it if it can really help. If you can make sure it would be useful (see @dannye's comments above), then I could give it a try.

In this case, please let me know how to remove the delay. Please post a patch so we are sure I'm doing the same that you did.

**Comment by ManoloFLTK** [comment]
@Albrecht-S Here is the patch to remove the delay. TIA.
```
diff --git a/src/Fl_win32.cxx b/src/Fl_win32.cxx
index 809fd0479..959f6f85f 100644
--- a/src/Fl_win32.cxx
+++ b/src/Fl_win32.cxx
@@ -1614,7 +1614,9 @@ static LRESULT CALLBACK WndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lPar
               window->user_data() != &Fl_WinAPI_Screen_Driver::transient_scale_display) {
             Fl_WinAPI_Window_Driver::data_for_resize_window_between_screens_.busy = true;
             Fl_WinAPI_Window_Driver::data_for_resize_window_between_screens_.screen = news;
-            Fl::add_timeout(1, Fl_WinAPI_Window_Driver::resize_after_screen_change, window);
+            // remove delay for resize after window changed display
+            //Fl::add_timeout(1, Fl_WinAPI_Window_Driver::resize_after_screen_change, window);
+            Fl_WinAPI_Window_Driver::resize_after_screen_change(window);
           }
           else if (!Fl_WinAPI_Window_Driver::data_for_resize_window_between_screens_.busy)
             wd->screen_num(news);
```

**Comment by dannye** [comment]
With the delay (pretty stable):

https://github.com/fltk/fltk/assets/5231310/36fa650c-255c-45cc-9c6c-7321cef7dbc7

Without the delay (many problems):

https://github.com/fltk/fltk/assets/5231310/e3f541b7-380f-4d8a-8804-c483bfade9e7

**Comment by ManoloFLTK** [comment]
Here is another modification that may be tried: the delay is shortened rather than completely removed.
```
diff --git a/src/Fl_win32.cxx b/src/Fl_win32.cxx
index 809fd0479..959f6f85f 100644
--- a/src/Fl_win32.cxx
+++ b/src/Fl_win32.cxx
@@ -1614,7 +1614,9 @@ static LRESULT CALLBACK WndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lPar
               window->user_data() != &Fl_WinAPI_Screen_Driver::transient_scale_display) {
             Fl_WinAPI_Window_Driver::data_for_resize_window_between_screens_.busy = true;
             Fl_WinAPI_Window_Driver::data_for_resize_window_between_screens_.screen = news;
-            Fl::add_timeout(1, Fl_WinAPI_Window_Driver::resize_after_screen_change, window);
+            // shorten delay for resize after window changed display
+            //Fl::add_timeout(1, Fl_WinAPI_Window_Driver::resize_after_screen_change, window);
+            Fl::add_timeout(0.01, Fl_WinAPI_Window_Driver::resize_after_screen_change, window);
           }
           else if (!Fl_WinAPI_Window_Driver::data_for_resize_window_between_screens_.busy)
             wd->screen_num(news);
```

**Comment by Albrecht-S** [comment]
@ManoloFLTK OK, I'll give it a try if I can start this old machine (it has some issues) but please don't hold your breath. Very likely not today.

**Comment by ManoloFLTK** [comment]
I have now understood the issue and am able to reproduce it here with the editor and the tree FLTK programs. I didn't know it was possible to maximize a window by moving it close to the display top. I also tested the keyboard-directed display change in my Windows 11 system which is a VM guest of a macOS host which I also ignored.

Overall, I find that with the change that reduces the delay to a very short time, FLTK windows move and rescale swiftly between displays, either dragged by hand or with Windows-Shift-left|right. But they don't behave correctly if they are simultaneously moved between screens with diverse scale factors and maximized by moving them to the display top.

I completely exclude to have Windows use the same scale factor on all its displays: that's key to support of displays with heterogeneous DPIs. 

If we can confirm that the reduced delay also works OK on earlier Windows version, I would propose that we may state that a known limitation of FLTK 1.4's Windows platform is that it's not always possible to simultaneously move a window to another display and maximize it by bringing it close to the display top. Notice that, after such a simultaneous operation, the window is damaged, but it's enough to unmaximize it to obtain a correct window [This doesn't remove the existence of this issue]. A better algorithm for rescaling maximized windows may hopefuly be implemented in a later version of FLTK.

**Comment by MatthiasWM** [comment]
My take on this:

A maximized window will simply lose its "maximized" status when zoomed. If a maximized window is zoomed up or down, it would first become unmaximized and it would then be scaled by whatever the maximized size was. If zoomed >100%, it would become bigger than the screen (assuming the window manager allows that). Aspect ratio, minimum size and maximum size don't need to be handled here at all.

If a fullscreen window is scaled, the default would be the same as above. There is only one exception. If the window is resizable with no aspect ratio, *and* the maximum *and* minimum size values are honored, the window should be zoomed and scaled simultaneously so it can remain at the same native size. If any of those fails, it loses its fullscreen attribute and back to step 1.

**Comment by dannye** [comment]
The shorter delay is an overall general improvement to the experience. It doesn't directly address any of the issues as far as I can tell, however.

> I would propose that we may state that a known limitation of FLTK 1.4's Windows platform is that it's not always possible to simultaneously move a window to another display and maximize it by bringing it close to the display top.

That isn't the only limitation though. It would be important to note that Shift+Win+Left/Right also do not always work. And it may likely affect Linux too (I just haven't tested yet), rather than just the Windows platform.

But respectfully, I really do not agree that this should be left as a "known limitation".

> I completely exclude to have Windows use the same scale factor on all its displays: that's key to support of displays with heterogeneous DPIs.

I understand your stance on this, but it really is the cause of all the issues AFAICT.

As a developer that uses FLTK, I would _much_ rather ship an application that requires the user to adjust the scaling of the window after moving it from one screen to another, rather than ship an application that corrupts the window instead. This would be unacceptable for anything that I ship myself.

I would also reiterate [a point I made previously](https://github.com/fltk/fltk/issues/919#issuecomment-1966627975) that didn't get any replies: This behavior also leads to a real inconvenience when using multiple _identical_ monitors:

> But now that you mention this, I do see the behavior you're talking about on Windows. I have noticed it in the past but thought it was a bug, because it was so disorienting when I'd drag a window from one screen to another and then its scale would suddenly change 1 full second later. This is kind of shocking for single-window applications IMO, especially when the two monitors have the same resolution.

And the [corresponding video](https://github.com/fltk/fltk/issues/919#issuecomment-1966655027).

This is a real annoyance that is only caused by each screen having its own scaling factor, despite the screens being totally identical.

---

I would also like to share another bug I just found after looking at the code in Fl_win32.cxx.

Since the `Fl::add_timeout()` callbacks are only serviced once the mouse button is release and the window is dropped, it's possible to stack multiple `Fl::add_timeout()` callbacks by sliding the window back and forth between two screens without letting go of the mouse button.

Depending on which screen the window is finally dropped on, its position and scale may be set incorrectly when the `Fl::add_timeout()` callbacks are finally serviced:

https://github.com/fltk/fltk/assets/5231310/f12eb3b6-88e3-4a22-8eb2-1bcd87248fda

**Comment by Albrecht-S** [comment]
@MatthiasWM wrote:
> A maximized window will simply lose its "maximized" status when zoomed. If a maximized window is zoomed up or down, it would first become unmaximized and it would then be scaled by whatever the maximized size was.

I second this in general, but if it's not clear: I would emphasize that the window would return back to its **original** size, and zooming would then apply as it would have done if the user had unmaximized the window. This would leave the window in its non-max (nor fullscreen) state but it's better than corrupting its contents. Users would get used to this behavior, I'm pretty sure.

**Comment by Albrecht-S** [comment]
@dannye wrote:
> Depending on which screen the window is finally dropped on, its position and scale may be set incorrectly when the Fl::add_timeout() callback is finally serviced: (video).

There's one thing that can be resolved easily: we can control that we don't queue multiple timeouts: simply use Fl::remove_timeout() before Fl::add_timeout(). Highlander motto: "There can be only one". 

The timeout itself does not use any scaling values (IIRC), hence it should be always correct if we call it related to the screen where the window is dropped. Unless I'm missing something.

**Comment by dannye** [comment]
I amended my comment to remove some incorrect assumptions I originally made about the exact cause of the bug. I added some printf's to prove that only 1 timeout is ever queued -- as soon as the window enters the new screen for the first time (they are not stacked).

I believe the fix would be similar to what you are saying though -- simply remove the timeout if the window is returned to its original screen before it is dropped.

**Comment by dannye** [comment]
> **_Edit:_** the timeout uses only the `window` as an argument, hence it should always be correct if we use only one timeout.

The implementation of `Fl_WinAPI_Window_Driver::resize_after_screen_change()` also uses `data_for_resize_window_between_screens_.screen` which is why it does the wrong thing even if called only once. `screen` refers to the wrong screen here.

**Comment by Albrecht-S** [comment]
@dannye Thanks for pointing this out. This sounds like the timeout should **retrieve** the correct data when it is executed rather than when it is scheduled. Good catch!

**Comment by dannye** [comment]
> I completely exclude to have Windows use the same scale factor on all its displays: that's key to support of displays with heterogeneous DPIs.

@ManoloFLTK @Albrecht-S Would it be possible to make this configurable to each FLTK user application? Expose an api to choose between "global" FLTK scaling or "per screen" FLTK scaling?

If it isn't too technically challenging, it could be the best compromise.

It could default to "per screen" scaling, and I (and others) could flip it to "global" scaling during initialization if we rather have no window corruption under any circumstances. Just an idea, curious to hear your thoughts.

**Comment by Albrecht-S** [comment]
My thought is: I know that it's useful, and it is also what Windows supports: per screen scaling - if the application allows it. Manolo put a lot of effort into it, and now where we are close to the release of FLTK 1.4.0 this is definitely not an option. All your tests and analysis are very much appreciated but this is not the time to discuss RFE's (requests for enhancement). Whether it is feasible to make this an option I can't tell, that's something Manolo would have to comment on. But, as I said, not right now (although he or any dev may add a comment, of course). However, we should strive to keep on topic, otherwise such an issue will never be completed.

If you want to suggest potential enhancements, please discuss this in fltk.coredev first. This is our current discussion forum for the development of FLTK (in the near future we may open GitHub discussions - or not). If it turns out that your suggestion would be feasible (and useful) then the next step would be to open a GitHub Issue with a sound description of the feature you request, and particularly with one issue (RFE) per feature. Thanks for your understanding.

**Comment by dannye** [comment]
Sure, if it's too big of an enhancement to use as the actual resolution to this issue, then I will give it a rest. I'm only wondering if such an enhancement might actually be easier/better than other possible fixes to the window corruption that this issue is about. (Not off topic at all.)

My only goal, as [you said yourself](https://github.com/fltk/fltk/issues/919#issuecomment-1965509075), is:

> My goal in this discussion is to make FLTK **_behave consistently_** and to **_avoid undefined behavior_**.

which I still currently see no way of achieving other than to allow "global" FLTK scaling, or to implement the rescaling of maximized/full screen windows which is already deferred to 1.5.0, or to implement Matthias' possible solution if that indeed has no major drawbacks.

**Comment by ManoloFLTK** [comment]
Commit 0eef561 has FLTK compare at display opening time the OS scaling factor values of all displays. If they are all equal (e.g., all 100%, all 150%) ctrl/+/-/0/ keystrokes will operate systemwide, that is all app windows on all displays will rescale. If they are not, GUI scaling remains as before, which is necessary to support computers mixing regular and HighDPI displays.

@Albrecht-S With this, computers with 2 identical displays no longer need to rescale an FLTK window when it gets moved between displays. Thus, no more time delay for such rescale operations. The usefulness to check what happens on old Windows versions with a time delay is much decreased.

**Comment by ManoloFLTK** [comment]
Commit 266647d should fix the issue with unequal DPIs for screens and moving a  window with the mouse first to another display and then back to starting display without releasing the mouse.

**Comment by Albrecht-S** [comment]
> The usefulness to check what happens on old Windows versions with a reduced time delay is much decreased.

Is it still necessary at all? As I wrote before it would be a time consuming process and I've been busy with CMake/MinGW/other stuff, hence I couldn't try it yet. Maybe I can change the system scaling factor of one display (although they are identical) - if that would help.

**Comment by ManoloFLTK** [comment]
With commit above, FLTK windows jump correctly between screens with distinct scale factors with shortcut Windows+Shift+arrow.

**Comment by dannye** [comment]
@ManoloFLTK Thank you for all the updates. Many aspects are greatly improved.

I do believe that eb3977a should be reverted.

The commit message says "Fix regression introduced at 41740ca for non resizable windows" but when a widget has a null `resizable_`, that doesn't mean that the window is non resizable, it means that it doesn't have a child widget that will automatically adopt the resizing. The window is still resizable even with a null `resizable_`.

The effect of commit eb3977a is essentially that my application can now be resized again while maximized or full screen (which is what 41740ca originally fixed) because the expression `if (top->resizable() && (top->fullscreen_active() || top->maximize_active()))` is not true and therefore it will not return early like it should.

I'm not sure what regression commit eb3977a is referring to. Even if a window is truly non resizable, wouldn't we still want to block rescaling actions while such a non resizable window is maximized or full screen?

Other than this, things are looking really really nice on Windows. Thank you very much for everything. I will test again quickly on Linux and Mac since it's been a few days since I've tested.

**Comment by dannye** [comment]
I do see that the doc comments for `Fl_Window::default_size_range()` says:
```
  -# If resizable() is NULL (this is the default) then the window cannot
    be resized and the resize border and max-size control will not be
    displayed for the window.
```
but I don't interpret this as a statement about `resizeable_` in general, just a statement about the effect of calling `default_size_range()`. ie, if `default_size_range()` is called and `resizable_` is NULL, then `default_size_range()` will make the window non resizable.

But in my application I do not call `default_size_range()` and I also do not want a resizable child widget. I explicitly call `size_range()` to set a min and max size (which implies the window is resizable), and I deliberately keep `resizable_` as null because I don't want any "stretchy" child widgets when the window is resized.

Hopefully I'm not misunderstanding anything, but please let me know if I am.

**Comment by Albrecht-S** [comment]
> I don't interpret this as a statement about resizeable_ in general, just a statement about the effect of calling default_size_range(). ie, if default_size_range() is called and resizable_ is NULL, then default_size_range() will make the window non resizable.

@dannye: This interpretation is correct.

> I do not call default_size_range() and I also do not want a resizable child widget. I explicitly call size_range() to set a min and max size (which implies the window is resizable), and I deliberately keep resizable_ as null

This is correct as well.

_Please excuse my nitpicking for clarification: you can't call default_size_range() anyway unless you derive your own subclass from Fl_Window but calling default_size_range() wouldn't make sense then. However, calling size_range() explicitly (before show(), hopefully) will inhibit default_size_range() being called internally. The net effect is that default_size_range() is not called, which is what you meant._

The documentation of resizable(Fl_Widget*) states:
> If the resizable is NULL the group's size is fixed and all of the widgets in the group remain a fixed size and distance from the top-left corner. This is the default for groups derived from Fl_Window and Fl_Pack.

AFAICT this is not 100% correct: "the group's size is fixed" would imply that a top level window - which is "a group" - with resizable(NULL) would not be resizable. This is not true: it can be made resizable by setting size_range() as you (@dannye) do. This part of the docs needs clarification.

@ManoloFLTK That said, I can't tell if the mentioned commit should be reverted as @dannye suggests or what the correct solution would be but I agree that having a top-level window with _resizable() == NULL_ does not automatically mean it's not resizable.

**Comment by ManoloFLTK** [comment]
The regression commit eb3977a is referring to is that a non-resizable window such as that of FLTK's hello program would respond only once to ctrl+. So something has to be done. But I see now the fix at eb3977a has problems. I will investigate.

**Comment by ManoloFLTK** [comment]
7fea97c hopefully fixes correctly the regression about zooming non resizable windows.

**Comment by dannye** [comment]
> The regression commit [eb3977a](https://github.com/fltk/fltk/commit/eb3977ac4329aad3a09b88a1c3ff96284c47ac33) is referring to is that a non-resizable window such as that of FLTK's hello program would respond only once to ctrl+.

Hm, I don't quite understand.

A non-resizable window (a window whose min size is equal to its max size) can still be full screened (can have a button or shortcut to enter full screen mode and have a resize() handler to support it).

In such a scenario, 7fea97c provides another way to incorrectly rescale a presently full screen window.

This is what I meant by:

> Even if a window is truly non resizable, wouldn't we still want to block rescaling actions while such a non resizable window is maximized or full screen?

I'm curious what you mean by "would respond only once". Do you mean that a non-resizable window would scale up once when ctrl+ is pressed, but on subsequent presses of ctrl+ the window is detected as either maximized or full screen and therefore `Fl_Screen_Driver::scale_handler()` returns early when it shouldn't?

If so, I think the underlying bug is that the non-resizable window is erroneously getting marked as either maximized or full screen after being scaled up once.

**Comment by ManoloFLTK** [comment]
> The effect of commit eb3977a is essentially that my application can now be resized again while maximized or full screen.

Is that still true? If yes, please describe exactly a scenario showing the problem.  If no, please consider whether this issue can be closed.

**Comment by dannye** [comment]
I have moved on from that issue. I am explaining the flaw with the fix in 7fea97c:

**A non-resizeable full screen window will now be incorrectly scaled while it is presently full screen** because `Fl_Screen_Driver::scale_handler()` will not return early (ie, `if (Fl_Window_Driver::driver(top)->is_resizable() && (top->fullscreen_active() || top->maximize_active()))` resolves to false, despite the window being full screen).

To see this for yourself, apply this patch to test/fullscreen.cxx:

```diff
diff --git a/test/fullscreen.cxx b/test/fullscreen.cxx
index 780eae32d..a357ea3b3 100644
--- a/test/fullscreen.cxx
+++ b/test/fullscreen.cxx
@@ -298,6 +298,7 @@ int main(int argc, char **argv) {
     Fl::fatal("Options are:\n -2 = 2 windows\n -f = startup fullscreen\n%s",Fl::help);
 
   fullscreen_window window(460,400+30*NUMB); window.end();
+  window.size_range(460, 400 + 30 * NUMB, 460, 400 + 30 * NUMB);
 
   shape_window sw(10,10,window.w()-20,window.h()-30*NUMB-120);
   sw.set_visible(); // necessary because sw is not a child of window
```

Run fullscreen.exe, toggle full screen, and rescale the window up or down and see the bug.

Note that it is irrelevant that this fixed size of 460x670 is less than the size of the full screen window -- the constructor and `size_range()` could just as well have been called with the dimensions of the full screen.

I maintain that 7fea97c and eb3977a should be reverted. If there is some other bug where the rescale command "only responds once" then there must be a different underlying bug, which I cannot reproduce. For the rest of my thoughts on this issue please reread my previous comment. (The question in my previous comment was not rhetorical -- I would really like to better understand what you mean by "would respond only once".)

**Comment by ManoloFLTK** [comment]
@dannye  First, let me clarify what is the regression mentioned above:
- it occurs with linux and macOS, but not with Windows (that's probably why you didn't see it)
- return to 41740ca which creates the regression
- run FLTK's hello program under macOS for example, type ⌘+, the window zooms in, 
type ⌘+ again, the window doesn't change. Here's the problem.

Second, I understand now what is the issue you report. I confess I had never considered one would like to turn fullscreen a non-resizable window (except at window creation time).

The thing is we see 2 very different FLTK behaviors when `Fl_Window::fullscreen()` is applied to a non-resizable window, using the modified fullscreen program:
- under macOS, X11 and Wayland, the window doesn't change size;
- under Windows, the window turns fullscreen.

In both cases FLTK's internal FULLSCREEN flag for the window is turned ON which is not consistent under macOS, Wayland and X11.

At this point I would believe the Windows behavior is not correct, but I would like to understand what's your view about that. Would one expect a non resizable window which initially is not fullscreen to become fullscreen (or maximized)? For this to make sense, the window should have a resizable() somewhere so it redraws adapted to the fullscreen or maximized state. Overall, that makes a window with min size == max size so it's non-resizable and has no maximize button on its titlebar, but that one can put fullscreen or maximized and back to its min size programmatically. Why setting min size == max size  for that window in the first place? I don't understand. 

Clearly, FLTK should support as many scenarios as possible that a program may wish. The point here is to define a cross-platform behavior that FLTK will implement when `Fl_Window::fullscreen()` is applied to a mapped, non fullscreen, non-resizable window.

**Comment by Albrecht-S** [comment]
My take on this: 

1. If a window is not resizable then the **_user_** can't make it fullscreen because the fullscreen button is not available. This should also be the case when the program calls `Fl_Window::fullscreen().` According to Manolo this is already the case on all supported platforms but Windows, and therefore Windows should behave consistently. This is a bug on the Windows platform and should be fixed.

2. If any of the `fullscreen_active()` or `maximize_active()` flags are turned on although the window is not made fullscreen or maximized then this is a bug and should be fixed.

3. If an app programmer wants to make a window fullscreen by calling `Fl_Window::fullscreen()` then they must make the window resizable, otherwise point 1. would prevent this. This could be considered a regression (if you could make a non-resizable window fullscreen in 1.3) but I would call it a bug fix and not a regression, particularly if this false behavior only "worked" on Windows.

I believe that this more consistent cross-platform behavior would fix some of the observations made above - or at least it could simplify the code in some places. Example: the "fix" in 7fea97c7a7dd36ad15b8d70f339c901adf88ba46 where the first condition (is_resizable()) would become irrelevant and could be removed. The only condition would be the current window state.

**Comment by dannye** [comment]
> @dannye First, let me clarify what is the regression mentioned above:
> 
>   * it occurs with linux and macOS, but not with Windows (that's probably why you didn't see it)
>   * return to [41740ca](https://github.com/fltk/fltk/commit/41740caa2e8542c155980486638f3fb0b781862a) which creates the regression
>   * run FLTK's hello program under macOS for example, type ⌘+, the window zooms in,
>     type ⌘+ again, the window doesn't change. Here's the problem.

@ManoloFLTK OK, then I would reiterate:

> Do you mean that a non-resizable window would scale up once when ctrl+ is pressed, but on subsequent presses of ctrl+ the window is detected as either maximized or full screen and therefore `Fl_Screen_Driver::scale_handler()` returns early when it shouldn't?
> 
> If so, I think the underlying bug is that the non-resizable window is erroneously getting marked as either maximized or full screen after being scaled up once.

You said:

> I had never considered one would like to turn fullscreen a non-resizable window **_(except at window creation time)_**.

A non-resizable window which is full screen at window creation time is exactly what I meant by:

> (Edit: an absolutely-non-resizable, permanently-full-screen application would also be susceptible to this bug.)

which is also prone to the bug in 7fea97c. (Isn't this reason enough to say non-resizable full screen windows are totally valid?)

> The thing is we see 2 very different FLTK behaviors when `Fl_Window::fullscreen()` is applied to a non-resizable window, using the modified fullscreen program:
> 
>     * under macOS, X11 and Wayland, the window doesn't change size;
> 
>     * under Windows, the window turns fullscreen.
> 
> 
> In both cases FLTK's internal FULLSCREEN flag for the window is turned ON which is not consistent under macOS, Wayland and X11.
> 
> At this point I would believe the Windows behavior is not correct, but I would like to understand what's your view about that. Would one expect a non resizable window which initially is not fullscreen to become fullscreen (or maximized)?

It's definitely debatable, but I think the Windows behavior is correct/expected. The application programmer may want the window to support 1 windowed size plus full screen mode by implementing Fl_Window::resize() to support it.

At the very least, a fixed-size non-resizable window should be full screenable if that fixed size matches the full screen size.

> For this to make sense, the window should have a resizable() somewhere so it redraws adapted to the fullscreen or maximized state.

No, I think "should" is too strong of a word here. As usual, a null resizable_ should be totally fine imo. "Expecting" a resizable_ I think is making too many assumptions about the application and its potential Fl_Window::resize() override.

> Overall, that makes a window with min size == max size so it's non-resizable and has no maximize button on its titlebar, but that one can put fullscreen or maximized and back to its min size programmatically. Why setting min size == max size for that window in the first place? I don't understand.

Video games do it all the time. Instead of "why would someone do it?", I think the appropriate question is "should it be valid?" and I think the answer is clearly "yes".

@Albrecht-S said:

>   1. If a window is not resizable then the **_user_** can't make it fullscreen because the fullscreen button is not available. This should also be the case when the program calls `Fl_Window::fullscreen().` According to Manolo this is already the case on all supported platforms but Windows, and therefore Windows should behave consistently. This is a bug on the Windows platform and should be fixed.

I kind of disagree, for reasons like I've stated already.

Mainly that a windowed window may have a fixed size which matches a full screen window, in which case going full screen does not require a resize at all. This is what I meant by:

> Note that it is irrelevant that this fixed size of 460x670 is less than the size of the full screen window -- the constructor and `size_range()` could just as well have been called with the dimensions of the full screen.

I think there are valid reasons to want to make a non-resizable window capable of switching between full screen mode and windowed mode.

>   2. If any of the `fullscreen_active()` or `maximize_active()` flags are turned on although the window is not made fullscreen or maximized then this is a bug and should be fixed.

I think this is the real cause of the "only responds to ctrl+ once" bug.

>   3. If an app programmer wants to make a window fullscreen by calling `Fl_Window::fullscreen()` then they must make the window resizable, otherwise point 1. would prevent this. This could be considered a regression (if you could make a non-resizable window fullscreen in 1.3) but I would call it a bug fix and not a regression, particularly if this false behavior only "worked" on Windows.

That could be a reasonable compromise in order to keep FLTK and the user application in full agreement about what is and is not valid.

> I believe that this more consistent cross-platform behavior would fix some of the observations made above - or at least it could simplify the code in some places. Example: the "fix" in [7fea97c](https://github.com/fltk/fltk/commit/7fea97c7a7dd36ad15b8d70f339c901adf88ba46) where the first condition (is_resizable()) would become irrelevant and could be removed. The only condition would be the current window state.

I think that would be best.

Another idea (for future FLTK) is that windows could have a dedicated "fullscreenable" flag, and if true then the window could be made full screen even if doing so would make it go beyond the current max size range. Maybe complicated and maybe not too useful, and certainly off topic from this issue, but just an idea for the future.

**Comment by dannye** [comment]
I re-tested on Linux and noticed a few issues.

Firstly, I noticed a bug that is related to this issue but different enough that I submitted the bug as #923.

Second, I noticed that the "full second delay" when moving a scaled window from one screen to another (which was reduced to 0.01 seconds on Windows) also applies to Linux, as well as the inconvenient behavior where multiple identical displays have separate distinct scaling factors.

Can these improvements already done on Windows also now be applied to Linux as well? Thanks.

(For the sake of my Linux testing, my Linux setup uses three 1920x1080 displays all with a DPI of 93.6x93.6 according to test/fullscreen.)

**Comment by dannye** [comment]
Related to the full second delay, and also related to the previous issue of the Fl::add_timeout while dragging the window back and forth without dropping the window, I noticed this rather amusing bug:

https://github.com/fltk/fltk/assets/5231310/172872a8-a87a-42cd-a038-d7bb41f707ae

**Comment by Albrecht-S** [comment]
@dannye Please tell us what you did and what you expected when you post a video. Without such a description we can't tell what the "amusing" part of this video shall be. Please describe (not only now, but also in the future). Note that we devs may not have the same or a similar configuration like you have (multiple screen setup) and thus need more input. TIA.

**Comment by dannye** [comment]
Did you watch the video? It genuinely speaks for itself. But here are the steps:

1. Launch test/fullscreen with 2 or more displays
2. Rescale the window on at least one display so that the scaling is sufficiently different between the two displays (in the video I scaled down on the left display and scaled up on the right display)
3. Drag the window from one screen to the other and then back to the first screen and then let go

The app will be stuck in an infinite loop of rescaling and resizing, being sent back and forth from one screen to the other.

**Comment by ManoloFLTK** [comment]
Commit 130f5b6 should fix the issue around interactions between non-resizable, fullscreen, maximized, and rescale shortcuts:
- a non-resizable toplevel window (that is, with min size == max size) will not turn fullscreen nor be maximized whatever the OS;
- a fullscreen or maximized window blocks zoom shortcuts from being effective;
- to make a window fullscreen or maximized already at creation time, do'nt make it non-resizable and call fullscreen() or maximize() before show()'ing it.

**Comment by ManoloFLTK** [comment]
>   - Launch test/fullscreen with 2 or more displays
    - Rescale the window on at least one display so that the scaling is sufficiently different between the two displays (in the video I scaled down on the left display and scaled up on the right display)
    - Drag the window from one screen to the other and then back to the first screen and then let go

I do and redo that here and can't reproduce the problem: no window being moved endlessly between displays. I tried both FLTK apps in Wayland and X11 modes.

**Comment by dannye** [comment]
I pulled from FLTK master and removed my custom changes to fullscreen.cxx to ensure it is still happening:

https://github.com/fltk/fltk/assets/5231310/bc4ec0f9-c75d-4bac-b0ff-815b833f7373

It helps to make the window wider and to drop the window near the boundary between the two displays.

I believe this is a result of the fact that a handful of the improvements and bug fixes that have been made for Windows for this issue ought to also be applied to Linux.

Specifically 0eef561, 266647d, and 34b8801 (266647d being the relevant one for this effect I'm guessing).

(And I'll note that a quick google search indicates that some Linux distros do have a Shift+Super+Arrow keys shortcut just like Windows to move windows between displays but that does not appear to be present on my Linux machine so I cannot test it. But that is why I mention 34b8801 as well.)

If these improvements to Windows are able to be made consistent on Linux, I think this issue can be closed.

**Comment by ManoloFLTK** [comment]
I still cannot reproduce here the scenario ending in a window moving endlessly from screen to screen. My setting here (Debian or Ubuntu as VM guests of a macOS host) doesn't allow me to use a pure X11 desktop with multiple displays (it accepts only a single display). So I use a Wayland desktop (a.k.a., Mutter) with 2 displays and run the FLTK fullscreen app as an X11 client. Anyway commit 5ba756e extends to X11 the Windows fix about moving a window to a new display and then back without releasing the mouse. Does that remove the issue?

**Comment by ManoloFLTK** [comment]
About modifying the X11 FLTK platform to rescale all displays together like it was done for the Windows platform:

I believe a key difference between how X11 and Windows support multiple displays possibly mixing High and Low DPIs prevents from doing that. 
- Windows provides an API to read the OS scaling factor of each display (e.g., 100%, 150%, 200%,…) so it was enough to compare these values across displays and to have FLTK rescale all displays together when they all share the same factor at app start time.
- X11 provides a single piece of data, the value of the Xft.dpi X resource even for multiple, DPI-heteregenous displays; FLTK divides the value of this resource by 96 and gets the initial value by which FLTK drawing units are multiplied to obtain display pixel units. X11 gives access to the true pixel sizes of all displays, but not, to the best of my knowledge, to their real DPI or physical sizes in mm.  X11 does gives DPI and mm values, but, for some obscure reasons, they are plain wrong. So it's not possible, I believe, to determine reliably using X11 only whether screens are all low or all high DPI. That's far from ideal, but that's a consequence of X11 being much older than HighDPI displays.

In conclusion, I intend to let X11 rescale each of its display independently to preserve how FLTK supports mixed DPI display configurations. The true way out, I believe, is to go Wayland.

**Comment by dannye** [comment]
> Anyway commit [5ba756e](https://github.com/fltk/fltk/commit/5ba756e60aa7b988399189c4df0eb72cdf0364c8) extends to X11 the Windows fix about moving a window to a new display and then back without releasing the mouse. Does that remove the issue?

Yes, that absolutely fixes it. Thank you very much.

I think the only remaining limitation is on Linux and Windows, when a window is dragged to a different display with a different scaling factor and immediately maximized by dragging and dropping against the top edge of the display (or fullscreening the window before the 1.0 second delay is up).

I still think reducing the delay from 1.0 seconds to 0.01 seconds is a big net positive, but I realized that you didn't make that change to master. If you suspect this may be too risky of a change, I understand. But if there is no known risk, I think it should be a welcomed improvement.

I also think that a stopgap such as forcefully un-maximizing or un-fullscreening the window in these scenarios (right when the 1.0 delay is up and the window is about to get rescaled) would be a nice compromise for the time being. I will of course leave that up to you.

Otherwise, I think this issue is good to close.

**Comment by ManoloFLTK** [comment]
Please, close issue if you agree. The remaining points you list have been already considered and found not going to get changed rapidly.
